### PR TITLE
feat: Rotatory Positional Encoding 

### PR DESCRIPTION
**Pull request by abourramouss**
this pr is for the implementation of RoPE 

#195

### CODE REVIEW

File: source/nn/positional-encoding.lisp [hunk 1]
```diff
@@ -1,6 +1,39 @@
...
+  (let* ((x (car inputs))
+         (shape (shape x))
+         (last-two (last shape 2))
+         (n (first last-two))
+         (d (second last-two)))
+    (dotimes (i 10)
+      (format t "i = ~a~%" i)
+      (defparameter shape (shape x))
+      (defparameter x (!reshape x (list n d)))
+      (defparameter positions (range 0 n )) ;not sure if this is what mlx doing (mx.arange(n))
```
Comment (hikettei) [inline source/nn/positional-encoding.lisp:25]: you can use
```(!index-components (list shape))```
instead of `range`.

File: source/nn/positional-encoding.lisp [hunk 2]
```diff
@@ -1,6 +1,39 @@
...
 
+(defmodel (RoPE (dims &key (traditional NIL) (base 10000) (scale 1.0)))
+  ((dims dims)
+   (traditional traditional)
+   (base base)
+   (scale scale)))
+
+;RoPE needs an extra arg in the call function, the generic doesn't consider it
+;offset: int = 0
+(defmethod call ((op RoPE) &rest inputs)
```
Comment (hikettei) [inline source/nn/positional-encoding.lisp:15]: `:offset` should passed as an argument of RoPE model, not an argument for `call`

**Comment by abourramouss** [comment]
```
test-rope
  dtype=FLOAT32
    ✓ Shapes match
    ✓ Satisfying (atol=1.9999694) <= 2
    ✓ Satisfying (rtol=2.4570446) <= 3
T
NIL
```

Ideally the error should be on the e-5 scale, the issue is that the tensors aren't identical, some values are different.

File: source/nn/positional-encoding.lisp [hunk 3]
```diff
@@ -1,6 +1,50 @@
...
+         (positions (!index-components (list n)))
+         (freqs (!exp (!div (!mul (!index-components (list (floor d 2)))
+                                  (!const x (- (log 10000))))
+                            (!const x d))))
+         (positions-reshaped (!reshape positions (list n 1)))    ; Shape: (n,1)
+         (freqs-reshaped (!reshape freqs (list 1 (floor d 2))))  ; Shape: (1, D/2)
+         (theta (!mul positions-reshaped freqs-reshaped))        ; Shape: (n, D/2)
+         (costheta (!cos theta))
+         (sintheta (!sin theta))
+         (x1 (!view x 't 't `(0 ,d 2)))  ; x[..., 0:d:2]
```
Comment (hikettei) [inline source/nn/positional-encoding.lisp:35]: t instead of 't. (the symbol t is preserved by cl so actually you do not have to quote it)

**Comment by hikettei** [comment]
Your code looks very clean! LGTM
With regard to the accuracy issue, let me make some experiments (yeah it should be fit in 1e-5 scale) this might be due to the compiler issue

File: source/test-suite/test-rope.lisp [hunk 4]
```diff
@@ -0,0 +1,42 @@
+(in-package :caten/test-suite)
+
+(defun !rope (x)
```
Comment (hikettei) [inline source/test-suite/test-rope.lisp:3]: you wanna have this function in caten/nn?

**Comment by hikettei** [comment]
will merge this after fixing the error in 1e-5 scale

File: source/test-suite/test-rope.lisp [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    return x_out.type_as(x)")
+
+
+(import-function "torch_rope")
+
+(deftest test-rope
+  (with-given-dtype ((:float32 . "float32"))
+    (let ((x (rand `(1 20 20 20))))
+      (assert-equal
+       (:atol 2 :rtol 3)
```
Comment (hikettei) [inline source/test-suite/test-rope.lisp:34]: can you update this line to atol=1e-5, rtol=1e-5? i'd like to see the result on ci

File: source/nn/positional-encoding.lisp [hunk 6]
```diff
@@ -1,6 +1,50 @@
...
+         (rx2 (!add (!mul x1 sintheta) (!mul x2 costheta)))
+         (rx1-expanded (!reshape rx1 (append (shape rx1) (list 1))))
+         (rx2-expanded (!reshape rx2 (append (shape rx2) (list 1))))
+         (result (!concatenate -1 rx1-expanded rx2-expanded))    ; Shape: (b, n, half-dim, 2)
+         (rotated (!reshape result (list b n (* 2 (floor d 2)))))
+         (final-result (if (evenp d)
+                           rotated
+                           (let ((last-elem (!view x 't 't `(,(- d 1) ,d))))  ; Shape: (b, n, 1)
+                             (!concatenate -1 rotated last-elem)))))
+    (proceed (!reshape final-result shape))))
```
Comment (hikettei) [inline source/nn/positional-encoding.lisp:50]: remove the `proceed` from the `call` (that might why jit fails)

**Comment by hikettei** [comment]
Hmm with JIT it should be scheduled to a single kernel? (don't worry about that I will take this!)
```
CATEN/TEST-SUITE> (call (Rope `(10)) (make-tensor `(10 10)))
{Tensor[float32] :shape (10 10) :id TID1450187
  :buffer nil
  :op #<RESHAPE {700BF75D83}>
  :requires-grad NIL
  :variables (TID1450176)
  :tracker #<TRACKER :order={row(0 1)} :shape=(10 10) :contiguous-p=T>}
CATEN/TEST-SUITE> (proceed *)
[graph-schedule] Schedule Graph:

FastGraph[seen=NIL, outputs=(val_39)] {
    { Allocate } : [ val_34 <- (1 10 5 2) where lowered-p=nil ]
    { Allocate } : [ val_32 <- (1 10 5) where lowered-p=nil ]
    {  KERNEL  } : [ val_35 <- val_23, val_22, val_21, val_32, val_34 where lowered-p=nil :name=FUSED_CONCATENATENODE1463528]
    { Allocate } : [ val_0 <- (1 10 5) where lowered-p=nil ]
    { Allocate } : [ val_23 <- (10 10) where lowered-p=nil ]
    { Allocate } : [ val_17 <- (10 5) where lowered-p=nil ]
    { Allocate } : [ val_15 <- (10) where lowered-p=nil ]
    { Allocate } : [ val_9 <- (5) where lowered-p=nil ]
    { Allocate } : [ val_7 <- NIL where lowered-p=nil ]
    { Allocate } : [ val_5 <- NIL where lowered-p=nil ]
    { Allocate } : [ val_3 <- NIL where lowered-p=nil ]
    { Allocate } : [ val_1 <- NIL where lowered-p=nil ]
    {  KERNEL  } : [ val_22, val_21, val_36 <- val_1, val_3, val_5, val_7, val_9, val_15, val_17, val_23, val_0, val_35 where lowered-p=nil :name=FUSED_CONCATENATENODE_COSNODE1463526]
    { Allocate } : [ val_37 <- (1 10 5 2) where lowered-p=nil ]
    {  KERNEL  } : [ val_38 <- val_37, val_36 where lowered-p=nil :name=FUSED_MOVE1463522]
    {   VMOP   } : [ val_39 <- val_38 where lowered-p=nil :name=FUSED_BACKWARD1463520]
}

[14:21:53, 11/20/2024 (GMT+9)] : JIT Compilation Start (AVM=MAIN1450194)

* (1/3) FUSED_CONCATENATENODE1463528
=====> Lowering to blueprint
{
  for (int _gid0=0;(_gid0<1);_gid0+=1) {
    for (int _gid1=0;(_gid1<10);_gid1+=1) {
      for (int _gid2=0;(_gid2<5);_gid2+=1) {
        val_27 = -((val_23[((_gid0+(10*_gid1))+(1+(2*_gid2)))]*val_22[((_gid0+(5*_gid1))+_gid2)]));
        val_35[(((100*_gid0)+(10*_gid1))+(2*_gid2))] = ((val_23[((_gid0+(10*_gid1))+(2*_gid2))]*val_21[((_gid0+(5*_gid1))+_gid2)])+val_27);
      } // _gid2
    } // _gid1
  } // _gid0
}
Compilation Time : 0.030271(sec)
* (2/3) FUSED_CONCATENATENODE_COSNODE1463526
=====> Lowering to blueprint
{
  for (int _gid0=0;(_gid0<1);_gid0+=1) {
    for (int _gid1=0;(_gid1<10);_gid1+=1) {
      for (int _gid2=0;(_gid2<5);_gid2+=1) {
        val_19 = (_gid1*exp2((((_gid2*-9.2103405)*0.1)*1.442695)));
        val_22[((5*_gid1)+_gid2)] = sin(val_19);
        val_21[((5*_gid1)+_gid2)] = sin((val_19+1.5707964));
        val_29 = (val_23[((_gid0+(10*_gid1))+(1+(2*_gid2)))]*val_21[((_gid0+(5*_gid1))+_gid2)]);
        val_36[((((100*_gid0)+(10*_gid1))+(2*_gid2))+1)] = ((val_23[((_gid0+(10*_gid1))+(2*_gid2))]*val_22[((_gid0+(5*_gid1))+_gid2)])+val_29);
      } // _gid2
    } // _gid1
  } // _gid0
}
Compilation Time : 0.04597(sec)
* (3/3) FUSED_MOVE1463522 

=====> Lowering to blueprint
{
  for (int _gid0=0;(_gid0<100);_gid0+=1) {
    val_38[_gid0] = val_36[_gid0];
  } // _gid0
}
Compilation Time : 0.001554(sec)
```

File: source/test-suite/test-rope.lisp [hunk 7]
```diff
@@ -0,0 +1,42 @@
+(in-package :caten/test-suite)
```
Comment (hikettei) [inline source/test-suite/test-rope.lisp:1]: If you create `test-rope.lisp`, you have to add this file into `caten/test-suite/caten.test-suite.asd`. otherwise lisp wont recognise this

File: source/nn/positional-encoding.lisp [hunk 8]
```diff
@@ -1,6 +1,50 @@
+
+
 (in-package :caten/nn)
 
-;; from https://github.com/ml-explore/mlx/blob/main/python/mlx/nn/layers/positional_encoding.py
+; from https://github.com/ml-explore/mlx/blob/main/python/mlx/nn/layers/positional_encoding.py
 ;; RoPE
 ;; PositionalEncoding
 
+(defmodel (RoPE (dims &key (traditional NIL) (base 10000) (scale 1.0) (offset 1.0)))
```
Comment (hikettei) [inline source/nn/positional-encoding.lisp:9]: dims is not used right?

**Comment by hikettei** [comment]
#245 will it fix something for JIT=0?

**Comment by abourramouss** [comment]
Been doing some tests in torch and mlx, the issue is that mlx and pytorch yield different results, and my implementation is basically inspired by mlx, while the test is using the torch implementation

This is the code i've been using:

```
import torch
from torchtune.modules import RotaryPositionalEmbeddings
from mlx.nn import RoPE
import mlx.core as mx
import numpy as np

# Parameters
seq_len, num_heads, head_dim = 30, 30, 30
dim = head_dim
max_seq_len = seq_len
theta = 10000.0

x = torch.rand(1, seq_len, num_heads, head_dim)
rope_torch = RotaryPositionalEmbeddings(dim=head_dim)

rope_mlx = RoPE(dims=head_dim)

numpy_array = x.detach().cpu().numpy()

x_mlx = mx.array(numpy_array)

mx_output = rope_mlx(x_mlx)

print("Output MLX:")
mlx_tensor = torch.tensor(numpy_array)
print(type(mlx_tensor))

print("Output PyTorch:")
torch_tensor = rope_torch(x)
print(type(torch_tensor))


absolute_diff = torch.sum(torch.abs(torch_tensor - mlx_tensor))

relative_diff = absolute_diff / torch.sum(torch.abs(mlx_tensor))

# Output results
print(f"Total Absolute Difference: {absolute_diff.item()}")
print(f"Total Relative Difference: {relative_diff.item()}")
```

and rtol atol:

```
Output MLX:
<class 'torch.Tensor'>
Output PyTorch:
<class 'torch.Tensor'>
Total Absolute Difference: 6711.056640625
Total Relative Difference: 0.49696406722068787
```

I will reimplement the call function using the torch implementation in order to lower the rtol and atol to the required values.

File: source/nn/positional-encoding.lisp [hunk 9]
```diff
@@ -1,6 +1,58 @@
...
+  (with-slots (dim base) op
+    (let* ((x (first inputs))
+           (x-shape (shape x))  ; x-shape: (batch-size, seq-len, num-heads, head-dim)
+           (batch-size (nth 0 x-shape))
+           (seq-len (nth 1 x-shape))
+           (num-heads (nth 2 x-shape))
+           (head-dim (nth 3 x-shape)))
+      ;; Validate that dim matches head_dim / 2
+      (when (/= dim (/ head-dim 2))
+        (error "Mismatch: Provided dim (~A) does not match head-dim / 2 (~A)" dim (/ head-dim 2)))
```
Comment (hikettei) [inline source/nn/positional-encoding.lisp:20]: this line should be an assert
```lisp
(assert (/= dim (/ head-dim 2)) () "Mismatch: Provided dim (~A) does not match head-dim / 2 (~A)" dim (/ head-dim 2))
```

File: source/nn/positional-encoding.lisp [hunk 10]
```diff
@@ -1,6 +1,58 @@
...
+             (seq-idx-reshaped (!reshape seq-idx (list seq-len 1)))       ; Shape: (seq-len, 1)
+             (idx-theta (!mul seq-idx-reshaped theta-reshaped))           ; Shape: (seq-len, dim)
+             (cosine (!cos idx-theta))                                    ; Shape: (seq-len, dim)
+             (sine (!sin idx-theta))                                      ; Shape: (seq-len, dim)
+             (xshaped (!reshape x (list batch-size seq-len num-heads dim 2)))
+             (cosine-reshaped (!reshape cosine (list 1 seq-len 1 dim 1)))
+             (sine-reshaped (!reshape sine (list 1 seq-len 1 dim 1)))
+             (num-dimensions 5)
+             (x0-subscripts (make-list num-dimensions :initial-element 'T))
+             (x1-subscripts (make-list num-dimensions :initial-element 'T)))
```
Comment (hikettei) [inline source/nn/positional-encoding.lisp:37]: just i think it is more preferable to use t instead of 't
```
(make-list num-dimensions :initial-element t)
```

File: source/nn/positional-encoding.lisp [hunk 11]
```diff
@@ -1,6 +1,58 @@
...
+     (base base)))
+
+(defmethod call ((op RoPE) &rest inputs)
+  (with-slots (dim base) op
+    (let* ((x (first inputs))
+           (x-shape (shape x))  ; x-shape: (batch-size, seq-len, num-heads, head-dim)
+           (batch-size (nth 0 x-shape))
+           (seq-len (nth 1 x-shape))
+           (num-heads (nth 2 x-shape))
+           (head-dim (nth 3 x-shape)))
```
Comment (hikettei) [inline source/nn/positional-encoding.lisp:17]: wanna assert?
```lisp
(assert (= (ndim x) 4))
```

Comment (hikettei) [reply source/nn/positional-encoding.lisp:17]: (Optional) Once you assert the rank you can simplify the code like:
```lisp
(multiple-value-bind (batch-size seq-len num-heads head-dim) (apply #'values (shape x))
  ;; body
  ;; ...
)
```

File: source/nn/positional-encoding.lisp [hunk 12]
```diff
@@ -1,6 +1,58 @@
 (in-package :caten/nn)
 
-;; from https://github.com/ml-explore/mlx/blob/main/python/mlx/nn/layers/positional_encoding.py
 ;; RoPE
 ;; PositionalEncoding
 
+(defmodel (RoPE (dim &key (base 10000)))
+    ((dim dim)
```
Comment (hikettei) [inline source/nn/positional-encoding.lisp:7]: FYI: `(dim dim :type list) (base base :type fixnum)` to add a type
